bpf: service loopback for ipv6

saiaunghlyanhtet opened the pull request:
Service Loopback IPv6

Fixes: #26733 

```release-note
IPv6 support for pods connecting to themselves via a k8s service ("service loopback").
```

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1122,12 +1157,41 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
+			union v6addr loopback_addr = IPV6_LOOPBACK;
+			ipv6_addr_copy(&new_saddr, &loopback_addr);
+			state->loopback = 1;
+		#endif
```
I don't think this will pass the compiler. Note that you can extend `bpf/tests/tc_nodeport_test.c` as needed to iterate faster.

julianwiedmann commented on bpf/lib/lb.h:
```diff
@@ -1122,12 +1157,41 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
+	    saddr.addr[12] == backend->address.addr[12] &&
+	    saddr.addr[13] == backend->address.addr[13] &&
+	    saddr.addr[14] == backend->address.addr[14] &&
+	    saddr.addr[15] == backend->address.addr[15]) {
```
Ideally this would be `ipv6_addr_equals()`.

julianwiedmann commented on pkg/node/local_node_store.go:
```diff
@@ -39,6 +39,7 @@ type LocalNode struct {
 	// v6 CIDR in which pod IPs are routable
 	IPv6NativeRoutingCIDR *cidr.CIDR
 	IPv4Loopback          net.IP
+	IPv6Loopback          net.IP
```
Have you already looked at how this ends up in the datapath? You can follow how `router_ipv6` is plumbed in via `bpf/include/bpf/config/node.h` and used as `CONFIG()` in the datapath.

saiaunghlyanhtet replied:
@julianwiedmann 
Can I integrate IPv6Loopback like router_ipv6 via node.h? And if so, do I need to follow this docs https://github.com/cilium/cilium/blob/main/Documentation/contributing/development/datapath_config.rst , and run this command (make -C bpf -j$(nproc)) to auto-generate GO code in config?

julianwiedmann replied:
yep that looks like a great guide to follow! Please shout on Slack if anything in there is not clear, this part of the loader infrastructure is fairly recent.

saiaunghlyanhtet replied:
Can you have a look at the current condition of the PR if you have time? Anything else needs to be done apart from tc_nodeport_test.c?

julianwiedmann replied:
Thank you! I can't do detailed pre-review of PRs, sorry. More than enough on my plate :).

I'd suggest splitting the changes into two patches (one that introduces `IPV6_LOOPBACK` in the controlplane, and a second patch that then adds all the datapath parts). And ideally we'd have both a connectivity test and a BPF-level for this specific behavior in the service translation, to confirm that your changes work as expected.

Note that https://github.com/cilium/cilium/pull/38818 is making similar changes for IPv4, maybe you can cross-check whether that matches up and check with @jrife for review :).

jrife commented on bpf/include/bpf/config/node.h:
```diff
@@ -16,3 +16,4 @@
 #include <node_config.h>
 
 NODE_CONFIG(union v6addr, router_ipv6, "Internal IPv6 router address assigned to the cilium_host interface")
+NODE_CONFIG(union v6addr, ipv6_loopback, "IPv6 loopback address")
```
It looks like we ended up settling on `service_loopback_ipv4` in https://github.com/cilium/cilium/pull/38818. Rename this and follow the [same patterns](https://github.com/cilium/cilium/pull/38818/commits/3d057129899afe3b6e0e75743116f0cd4c0183da).

saiaunghlyanhtet replied:
Changed to service_loopback_ipv6.

jrife commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -278,6 +278,7 @@ cilium-agent [flags]
       --ipv6-node string                                          IPv6 address of node (default "auto")
       --ipv6-pod-subnets strings                                  List of IPv6 pod subnets to preconfigure for encryption
       --ipv6-range string                                         Per-node IPv6 endpoint prefix, e.g. fd02:1:1::/96 (default "auto")
+      --ipv6-service-loopback-address string                      IPv6 address for service loopback SNAT when a Pod talks to itself over a Service. (default "fd00::1")
```
nit: The language here slightly differs from `ipv4-service-loopback-address`. Keep them consistent.

#### Current
```
--ipv4-service-loopback-address string                      IPv4 source address to use for SNAT when a Pod talks to itself over a Service. (default "169.254.42.1")
...
--ipv6-service-loopback-address string                      IPv6 address for service loopback SNAT when a Pod talks to itself over a Service. (default "fd00::1")
...
```

#### Should Be
```
--ipv4-service-loopback-address string                      IPv4 source address to use for SNAT when a Pod talks to itself over a Service. (default "169.254.42.1")
...
--ipv6-service-loopback-address string                      IPv6 source address to use for SNAT when a Pod talks to itself over a Service. (default "fd00::1")
...
```

jrife commented on bpf/bpf_lxc.c:
```diff
@@ -543,6 +544,9 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		}
 #endif /* ENABLE_L7_LB */
 
+		if (hairpin_flow)
```
Can we add a comment here similar to the one in `handle_ipv4_from_lxc`?

jrife commented on bpf/bpf_lxc.c:
```diff
@@ -690,13 +696,23 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	if (is_defined(ENABLE_ROUTING) || is_defined(ENABLE_HOST_ROUTING)) {
```
I think this condition needs an additional `|| hairpin_flow` condition similar to the IPv4 logic. Have you tested with endpoint routes?

saiaunghlyanhtet replied:
Thank you for pointing that out.

jrife commented on bpf/bpf_lxc.c:
```diff
@@ -690,13 +696,23 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
+		 * for the destination endpoint lookup.
 		 */
-		ep = lookup_ip6_endpoint(ip6);
+		if (ct_status == CT_REPLY && hairpin_flow)
```
nit: I would slightly prefer if we could keep the structure similar to the IPv4 version. In particular, by declaring `union v6addr daddr = ip6->daddr` in this scope and reassigning its value to `saddr` so that there's only one place `__lookup_ip6_endpoint` is called.

jrife commented on bpf/lib/lb.h:
```diff
@@ -570,9 +587,11 @@ lb6_lookup_rev_nat_entry(struct __ctx_buff *ctx __maybe_unused, __u16 index)
  * @arg tuple		tuple
  */
 static __always_inline int lb6_rev_nat(struct __ctx_buff *ctx, int l4_off, __u16 index,
+					   bool loopback,
```
nit: Indention looks weird here. This doesn't look aligned.

jrife commented on bpf/lib/lb.h:
```diff
@@ -1123,13 +1155,30 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
+	}
+
+	if (ipv6_addr_equals(&saddr, &backend->address)) {
+	#ifdef USE_LOOPBACK_LB
```
This can just be moved into the if block above instead of doing `if (ipv6_addr_equals(&saddr, &backend->address))` again. This would be more consistent with `lb4_local` as well.

jrife commented on bpf/lib/lb.h:
```diff
@@ -1123,13 +1155,30 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
+#ifdef USE_LOOPBACK_LB
+	if (!state->loopback)
+#endif
+
```
nit: get rid of the empty line.

jrife commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -545,3 +551,425 @@ int tc_drop_no_backend_check(const struct __ctx_buff *ctx)
+	pktgen__init(&builder, ctx);
+
+	l4 = pktgen__push_ipv6_tcp_packet(&builder,
+									  (__u8 *)src, (__u8 *)dst,
```
fix indention

jrife commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -545,3 +551,425 @@ int tc_drop_no_backend_check(const struct __ctx_buff *ctx)
+	pktgen__init(&builder, ctx);
+
+	l4 = pktgen__push_ipv6_tcp_packet(&builder,
+									  (__u8 *)src, (__u8 *)dst,
```
fix indention.

jrife commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -545,3 +551,425 @@ int tc_drop_no_backend_check(const struct __ctx_buff *ctx)
+	pktgen__init(&builder, ctx);
+
+	l4 = pktgen__push_ipv6_tcp_packet(&builder,
+									  (__u8 *)src, (__u8 *)dst,
```
fix indention.

jrife commented on pkg/datapath/loader/config.go:
```diff
@@ -19,5 +19,9 @@ func nodeConfig(lnc *datapath.LocalNodeConfiguration) config.Node {
 		node.RouterIPv6 = ([16]byte)(lnc.CiliumInternalIPv6.To16())
 	}
 
+	if lnc.ServiceLoopbackIPv6 != nil {
```
Group this together with 

```
if lnc.ServiceLoopbackIPv4 != nil {
		node.ServiceLoopbackIPv4 = [4]byte(lnc.ServiceLoopbackIPv4.To4())
	}
```

andreaterzolo commented on bpf/lib/lb.h:
```diff
@@ -847,24 +866,35 @@ lb6_select_backend_id(struct __ctx_buff *ctx __maybe_unused,
 
+		if (ipv6_store_saddr(ctx, (void *)new_saddr->addr, l3_off) < 0)
+			return DROP_WRITE_ERROR;
+		sum = csum_diff(old_saddr->addr, 16, new_saddr->addr, 16, 0);
```
I believe we need to provide the previous `sum` to have a correct computation of the final checksum
```suggestion
		sum = csum_diff(old_saddr->addr, 16, new_saddr->addr, 16, sum);
```

andreaterzolo replied:
A slightly more performant version could be
```c
{
	const union v6addr *new_dst = &backend->address;
	struct csum_offset csum_off = {};
	__be32 sum = 0;

	if (has_l4_header)
		csum_l4_offset_and_flags(nexthdr, &csum_off);

	if (ipv6_store_daddr(ctx, new_dst->addr, l3_off) < 0)
		return DROP_WRITE_ERROR;
#ifdef USE_LOOPBACK_LB
	if (new_saddr && (new_saddr->d1 || new_saddr->d2)) {
		cilium_dbg_lb(ctx, DBG_LB6_LOOPBACK_SNAT, old_saddr->p4, new_saddr->p4);

		if (ipv6_store_saddr(ctx, (void *)new_saddr->addr, l3_off) < 0)
			return DROP_WRITE_ERROR;
		sum = csum_diff(old_saddr->addr, 16, new_saddr->addr, 16, 0);
	}
#endif /* USE_LOOPBACK_LB */
	if (csum_off.offset) {
		sum = csum_diff(key->address.addr, 16, new_dst->addr, 16, sum);

		if (csum_l4_replace(ctx, l4_off, &csum_off, 0, sum,
				    BPF_F_PSEUDO_HDR) < 0)
			return DROP_CSUM_L4;
	}

	if (!has_l4_header)
		return CTX_ACT_OK;

	return lb_l4_xlate(ctx, nexthdr, l4_off, &csum_off, key->dport,
			   backend->port);
}
```

in this way we compute `sum = csum_diff(key->address.addr, 16, new_dst->addr, 16, sum);` only when we really need to rewrite the l4 checksum. WDYT?

saiaunghlyanhtet replied:
I think your suggestion looks better. But, we may need to be consistent with `lb4_xlate`. I'd love to get other reviewers' opinions on this.

jrife replied:
> But, we may need to be consistent with lb4_xlate

IMO we should keep things consistent with the v4 stuff until this is merged. Then, if there's an optimization to be made, it could be done in another PR that targets both v4 and v6 at the same time.

andreaterzolo commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -545,3 +551,428 @@ int tc_drop_no_backend_check(const struct __ctx_buff *ctx)
+
+    if (l4->dest != tcp_dst_one)
+	test_fatal("dst TCP port incorrect");
+
```
maybe we could assert against the l4 checksum here `if (l4->check != bpf_htons(...))` like in the ipv4 tests

andreaterzolo replied:
Same in all other tests below

saiaunghlyanhtet replied:
Thanks for pointing that out. I forgot to add check for checksum.

andreaterzolo commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -545,3 +551,428 @@ int tc_drop_no_backend_check(const struct __ctx_buff *ctx)
+    ct_entry = map_lookup_elem(get_ct_map6(&tuple), &tuple);
+    if (!ct_entry)
+	test_fatal("no CT_SERVICE entry found");
+
```
We could also assert against the second conntrack entry created (`tuple.flags = TUPLE_F_OUT`), see the IPv4 corresponding test

andreaterzolo commented on bpf/lib/lb.h:
```diff
@@ -542,6 +543,22 @@ static __always_inline int __lb6_rev_nat(struct __ctx_buff *ctx, int l4_off,
+		ipv6_addr_copy(&old_daddr, &tuple->daddr);
+		cilium_dbg_lb(ctx, DBG_LB6_LOOPBACK_SNAT_REV, old_daddr.p4, old_saddr.p4);
+
+		ret = ipv6_store_daddr(ctx, old_saddr.addr, ETH_HLEN);
```
Shouldn't we call something like `csum_diff` to take the source address rewriting into consideration for the final L4_checksum update?

saiaunghlyanhtet replied:
I think we don't need to do that for IPv6 as we have already done L4 checksum update via `ipv6_l4_csum_update`.

andreaterzolo replied:
uhm, in the current code, we are recomputing the checksum only on the saddr change, but in the hairpin flow, we also change the daddr so we need to take it into account when we recompute the L4 checksum.

We can probably check it with the `hairpin_flow_4_reverse_ingress_v6` test.

I see you used this checksum `(l4->check != bpf_htons(0xdfd0))`, but if I put the packet generated after the SETUP phase into Wireshark 
```
00000000  13 37 13 37 13 37 de ad  be ef de ef 86 dd 60 00
00000010  00 00 00 28 06 40 fd 05  00 00 00 00 00 00 00 00
00000020  00 00 00 00 00 03 fd 04  00 00 00 00 00 00 00 00
00000030  00 00 00 00 00 01 00 50  57 3a 40 e2 01 00 00 00
00000040  00 00 50 12 ff ff df d0  00 00 53 68 6f 75 6c 64
00000050  20 6e 6f 74 20 63 68 61  6e 67 65 21 21 00
```
I see a validation error on the L4 checksum

```
Checksum: 0xdfd0 incorrect, should be 0xdfd1
```

This means that there is probably something wrong in the checksum computation, and I believe it is the fact that we are not recomputing it when we change the daddr

saiaunghlyanhtet replied:
Thanks. Fixed.

Andreagit97 reviewed (commented):
Hi! I'm not a maintainer or reviewer of this repo, but I was looking at this casually and noticed a few things. I've left some questions/comments about the L4 checksum computation. Hope they might be helpful!

jrife commented on bpf/include/bpf/config/node.h:
```diff
@@ -15,6 +15,7 @@
 /* Legacy node config rendered at agent runtime. */
 #include <node_config.h>
 
+NODE_CONFIG(union v6addr, service_loopback_ipv6, "IPv6 source address used for SNAT when a Pod talks to itself over a Service")
```
small nit: I feel like it makes more sense to have `service_loopback_ipv4` above `service_loopback_ipv6` here.

jrife commented on bpf/bpf_lxc.c:
```diff
@@ -697,16 +709,28 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
-		 *  - The destination IP address belongs to a local endpoint managed by
-		 *    cilium
-		 *  - The destination IP address is an IP address associated with the
-		 *    host itself.
```
nit: Remove these extra spaces that were introduced to make the diff cleaner.

jrife commented on bpf/bpf_lxc.c:
```diff
@@ -1639,6 +1665,19 @@ ipv6_policy(struct __ctx_buff *ctx, struct ipv6hdr *ip6, __u32 src_label,
 			break;
 
+#if defined(ENABLE_PER_PACKET_LB)
+		loopback_addr = CONFIG(service_loopback_ipv6);
```
Is `loopback_addr` necessary here? Can we just pass `&CONFIG(service_loopback_ipv6` directly to `ipv6_addr_equals`?

saiaunghlyanhtet replied:
I have already tried it. But, it leads to incompatible pointer type error.

jrife commented on bpf/lib/lb.h:
```diff
@@ -542,6 +543,22 @@ static __always_inline int __lb6_rev_nat(struct __ctx_buff *ctx, int l4_off,
+
+		/* Update the tuple address which is representing the destination address */
+		ipv6_addr_copy(&tuple->saddr, &old_saddr);
+    }
```
nit: this brace doesn't look aligned with the if statement above.

jrife commented on bpf/lib/lb.h:
```diff
@@ -579,9 +596,11 @@ lb6_lookup_rev_nat_entry(struct __ctx_buff *ctx __maybe_unused, __u16 index)
  * @arg ctx		packet
  * @arg l4_off		offset to L4
  * @arg index		reverse NAT index
+ * @arg loopback    loopback connection
```
nit: alignment with other args.

thorn3r reviewed (approved):
Approving for the agent-related work, it looks good :+1:  I think a more detailed commit message, and even splitting the datapath and agent changes into separate commits, would be nice

jrife commented on bpf/include/bpf/config/node.h:
```diff
@@ -15,12 +15,19 @@
-NODE_CONFIG(__u32, trace_payload_len, "Length of payload to capture when tracing native packets.")
+NODE_CONFIG(union v4addr, service_loopback_ipv4,
+	    "IPv4 source address used for SNAT when a Pod talks to itself over a Service")
+NODE_CONFIG(union v6addr, service_loopback_ipv6,
```
nit: checkpatch would probably be happy if you just made this adjustment to the lines you added. My preference would be to only change what you need for this PR and leave reformatting the other lines for another patch, but not a blocker.

saiaunghlyanhtet replied:
Fixed it. I think that the current checkpatch failed because of unknown manifest issue related t checkpatch image.

pchaigno replied:
Yes, it's unrelated to your changes. It will fix itself next time you rebase.

asauber commented on pkg/option/config.go:
```diff
@@ -773,6 +773,9 @@ const (
 	ServiceLoopbackIPv4 = "ipv4-service-loopback-address"
 
+	// ServiceLoopbackIPv6 is the address to use for service loopback SNAT
+	ServiceLoopbackIPv6 = "ipv6-service-loopback-address"
```
Sorry for the delay on this comment. We prefer that all new config options are added to a hive cell, to begin the process of modularizing the related feature. If you take a look at the [history on this file](https://github.com/cilium/cilium/commits/main/pkg/option/config.go), you will see some recent example of refactors into cells. The initial refactor can be purely for the sake of config, with other modularlization coming later. Could you look into making this change?

saiaunghlyanhtet replied:
To make it clear, I would like to confirm the cell file to which I need to move the current config. Is it ok to move to datapath/loader/cell.go? Or other file?

asauber replied:
Logically, pkg/ipam/cell/cell.go seems like the right place, but given the breadth of changes that are required in this PR, that might not be feasible on the dependency graph. Could you give that a try and see if you run into any issues?

@joamaki is there a good way to visualize the dependency graph in order to make this type of decision, or any other advice that you would provide here?

joamaki replied:
Since we already have `ServiceLoopbackIPv4` here I'd be OK with adding it here for now and then moving them both into a saner place in another PR. It'd probably make sense to do that as part of a bigger refactoring that moves `daemon/cmd/ipam.go` into `pkg/ipam`. @cilium/sig-ipam any opinion?

asauber reviewed (approved):
Approving for config. Given the above discussion, it seems like a separate refactor would be the place to move the flag.

pchaigno commented on pkg/defaults/defaults.go:
```diff
@@ -310,6 +310,9 @@ const (
 	ServiceLoopbackIPv4 = "169.254.42.1"
 
+	// ServiceLoopbackIPv6 is the default address for service loopback
+	ServiceLoopbackIPv6 = "fd00::1"
```
How was this selected? Wouldn't an IP address within the range reserved for link-local addresses (fe80::/64) be more appropriate?

saiaunghlyanhtet replied:
Just chose it because I think it is just private and routable within the cluster and interface-agnostic.

pchaigno replied:
Let's switch to a reserved link-local address then. Less chances of conflicts.

saiaunghlyanhtet replied:
Switched to LLA address.

pchaigno commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -537,3 +543,456 @@ int tc_drop_no_backend_check(const struct __ctx_buff *ctx)
+    data_end = (void *)(long)ctx->data_end;
+
+    if (data + sizeof(__u32) > data_end)
+	test_fatal("status code out of bounds");
```
There seem to be an indentation issue on all the `if` statements in this function.

saiaunghlyanhtet replied:
Fixed the indentation issue.

pchaigno commented on bpf/tests/tc_nodeport_test.c:
```diff
@@ -11,6 +11,7 @@
 
 /* Enable code paths under test*/
 #define ENABLE_IPV4
+#define ENABLE_IPV6
```
Could you move the new tests into their own commit?

Optionally: Probably even better if you can introduce them each in their own commit and describe in the commit descriptions what scenario is covered.

saiaunghlyanhtet replied:
I have moved the new tests into a single commit in which each test is properly described about what scenario is covered.

pchaigno reviewed (commented):
Thanks a lot for working on this! I looks really good :)

I only have minor comments below. IMO, it should be good to merge soon.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1583,6 +1607,7 @@ ipv6_policy(struct __ctx_buff *ctx, struct ipv6hdr *ip6, __u32 src_label,
 	__u8 policy_match_type = POLICY_MATCH_NONE;
 	__u8 audited = 0;
 	__u8 auth_type = 0;
+	union v6addr loopback_addr;
```
This variable can be unused so should be marked as `__maybe_unused`. This is why the CI is currently failing (or at least one reason).

saiaunghlyanhtet replied:
Thanks for pointing that out. Fixed it.

pchaigno commented on pkg/monitor/datapath_debug.go:
```diff
@@ -66,6 +66,8 @@ const (
 	DbgLb6ReverseNatLookup
 	DbgLb6ReverseNat
+	DbgLb6LoopbackSnat
+	DbgLb6LoopbackSnatRev
```
The integration tests are failing as a consequence of this change. You'll need to also update the list in `api/v1/flow/flow.proto` and regenerate the protobuf files, with `make generate-hubble-api`.

Ideally, I'd move the two new constants into their own commit and would note the command used to generate the files (i.e., `make generate-hubble-api`). Then, the new constants can be used in subsequent commits.

pchaigno replied:
Since you'll need to push, please rebase on latest `main` at the same time. That should help reduce the number of CI issues on the branch. I think we're close to being able to merge :slightly_smiling_face:

saiaunghlyanhtet replied:
Regenerated the flow.pb.go and rebased on the latest main.

pchaigno commented on api/v1/flow/flow.proto:
```diff
@@ -948,4 +948,6 @@ enum DebugEventType {
     DBG_L7_LB = 65;
     DBG_SKIP_POLICY = 66;
+    DBG_LB6_LOOPBACK_SNAT = 67;
+    DBG_LB6_LOOPBACK_SNAT_REV = 68;
```
This list needs to be in the same order as the list in `datapath_debug.go`, otherwise the constants won't get assigned the same values.

saiaunghlyanhtet replied:
I am sorry that I was rushing. Fixed it.

pchaigno replied:
No worries. Thanks for the quick replies :)

ti-mo reviewed (approved):
Thanks!

gandro commented on bpf/lib/dbg.h:
```diff
@@ -35,6 +35,8 @@ enum {
 	DBG_LB6_REVERSE_NAT_LOOKUP,
 	DBG_LB6_REVERSE_NAT,
+	DBG_LB6_LOOPBACK_SNAT,
+	DBG_LB6_LOOPBACK_SNAT_REV,
```
This is a bit of an unfortunate situation - but because we pull in these definitions in Hubble, they are considered stable. You have to move them to the end of this enum, to avoid re-numbering the existing enum variants. Otherwise older Hubble clients will now report the wrong trace type when decoding debug events (due to the changed integer-to-name mapping).

Long-term we probably have to find another solution, since I'm not sure it was a good idea to have this internal enum be part of the stable API. But I think that's something we can address separately later, independent of this PR. For this PR, I think it's easiest if we add this to the bottom of this enum (and accordingly for the Go and Hubble definitions).

pchaigno replied:
While at it, could you move the BPF definitions of the these two constants to the first commit, with the Go side? I think that will be a bit less confusing when reviewing :)

saiaunghlyanhtet replied:
@gandro Moved to the end of the enum.
@pchaigno Moved the dbg.h to the first commit.

gandro reviewed (changes requested):
Thanks for the PR. Unfortunately I think there are some minor changes needed

saiaunghlyanhtet committed:
api: hubble: regenerate flow.pb.go for IPv6 service loopback

- add DebugEventType DBG_LB6_LOOPBACK_SNAT, DBG_LB6_LOOPBACK_SNAT_REV
- regenerate flow.pb.go

Signed-off-by: saiaunghlyanhtet <saiaunghlyanhtet2003@gmail.com>

saiaunghlyanhtet committed:
cilium-agent: add IPv6 serviceloopback plumbing

Add agent-side support to enable IPv6 service loopback handling so that connections to a service’s own ClusterIP from a
backend pod loop back correctly.

Summary of changes:
- Extend service / LB logic to recognize and mark IPv6 loopback scenarios
- Introduce service_loopback_ipv6 runtime variable for IPv6 service loopback support

Signed-off-by: saiaunghlyanhtet <saiaunghlyanhtet2003@gmail.com>

saiaunghlyanhtet committed:
bpf: implement IPv6 service loopback datapath handling

Introduce datapath logic to support IPv6 service loopback so that packets originating from a pod targeting its own
service ClusterIP are correctly short-circuited and subject to the standard LB/rev-NAT path without external detours.
Adjust service lookup / backend selection and reverse NAT handling for IPv6 loopback.

Signed-off-by: saiaunghlyanhtet <saiaunghlyanhtet2003@gmail.com>

saiaunghlyanhtet committed:
bpf:test: Add BPF selftests to cover IPv6 loopback

hairpin_flow_1_forward_v6: Test that sending a packet from a pod to its own IPv6 service gets source NAT-ed
with the service loopback IPv6 address and is correctly forwarded to the pod's veth interface.
Verifies SNAT behavior and connection tracking entries for IPv6 hairpin traffic.

hairpin_flow_2_forward_ingress_v6: Test the ingress path for IPv6 hairpin flows where a packet from
the service loopback address to a pod creates the appropriate connection tracking entry on the ingress side
with the loopback flag.

hairpin_flow_3_reverse_v6: Test that return traffic in an IPv6 hairpin flow (from pod back to
service loopback address) is correctly handled and forwarded without any address translation,
maintaining the original flow semantics.

hairpin_flow_4_reverse_ingress_v6: Test the reverse direction ingress path for IPv6 hairpin flows where
return traffic gets NAT-ed back to the original service IP address, completing the hairpin loop and
ensuring the client sees responses from the expected service address.

tc_drop_no_backend_v6: Test that IPv6 packets destined to a service with no available backends
are correctly dropped with TC_ACT_SHOT, ensuring proper error handling for unavailable IPv6 services.

Signed-off-by: saiaunghlyanhtet <saiaunghlyanhtet2003@gmail.com>

gandro reviewed (approved):
Thanks!
